cpp/include/raft/distance/detail/matrix/matrix.hpp

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/error.hpp>
+
+namespace raft::distance::matrix::detail {
+
+template <typename math_t>
+class DenseMatrix;
+template <typename math_t>
+class CsrMatrix;
+
+/*
+ * Thin matrix wrapper to allow single API for different matrix representations
+ */
+template <typename math_t>
+class Matrix {
+ public:
+  Matrix(int rows, int cols) : n_rows(rows), n_cols(cols){};
+  virtual bool isDense() const = 0;
+  virtual ~Matrix(){};
+
+  DenseMatrix<math_t>* asDense()
+  {
+    DenseMatrix<math_t>* cast = dynamic_cast<DenseMatrix<math_t>*>(this);
+    ASSERT(cast != nullptr, "Invalid cast! Please check for isDense() before casting.");
+    return cast;
+  };
+
+  CsrMatrix<math_t>* asCsr()
+  {
+    CsrMatrix<math_t>* cast = dynamic_cast<CsrMatrix<math_t>*>(this);
+    ASSERT(cast != nullptr, "Invalid cast! Please check for isDense() before casting.");
+    return cast;
+  };
+
+  const DenseMatrix<math_t>* asDense() const
+  {
+    const DenseMatrix<math_t>* cast = dynamic_cast<const DenseMatrix<math_t>*>(this);
+    ASSERT(cast != nullptr, "Invalid cast! Please check for isDense() before casting.");
+    return cast;
+  };
+
+  const CsrMatrix<math_t>* asCsr() const
+  {
+    const CsrMatrix<math_t>* cast = dynamic_cast<const CsrMatrix<math_t>*>(this);
+    ASSERT(cast != nullptr, "Invalid cast! Please check for isDense() before casting.");
+    return cast;
+  };
+
+  int n_rows;
+  int n_cols;
+};
+
+template <typename math_t>
+class DenseMatrix : public Matrix<math_t> {
+ public:
+  DenseMatrix(math_t* data, int rows, int cols, bool row_major = false, int ld_in = 0)
```

Comment:
Ideally we would like to use mdspan for representing the dense matrix. Could you list the pain points that would motivate keeping a pointer based API still around?

Reply:
- I am fine keeping this version until the sparse wrapper PR #1279 is merged, and consolidate the interface afterwards.
- Note that we have helper functions to deal with mdspan on the Python side, see usage example here
https://github.com/rapidsai/raft/blob/03e26b520ea281d9b239938eda30dd5100559e24/python/pylibraft/pylibraft/neighbors/ivf_pq/ivf_pq.pyx#L523

Reply:
I agree with @tfeher, rather than introducing more abstractions for this, we can unify the dense and sparse APIs using overloads of the existing abstractions (with templates if needed).

We didn’t have abstractions like the mdspan available when we were building cuml so we used pointers for everything. Now that we have the proper abstractions and infrastructure in place, we should use them.

please note that the sparse API has been merged. We should be able to add overloads for both dense and sparse without having an explicit class to unify them.

Reply:
Tagging @divyegala for his thoughts here regarding public APIs and template metaprogramming techniques that might help.

Reply:
> rather than introducing more abstractions for this, we can unify the dense and sparse APIs using overloads of the existing abstractions (with templates if needed).

@cjnolet Having overloads instead of on unified matrix type might make sense when we consider this PR only. When looking at how this is used from cuML, that would motivate introducing a unified sparse dense type.

Let us consider the whole call chain in https://github.com/rapidsai/cuml/pull/5273:
- Cython warppers -> SvcFit -> SmoSolver.Solve -> KernelCache.getTile -> GramMatrix.evaluate
- Cython wrappers -> SvcPredict -> KernelCache.getTile -> GramMatrix.evaluate

Having overloads on this whole call chain is not feasible. Having the matrix type as template parameter would unnecessary add extra compilation steps when all we would need to pass on some metadata about the array structures.

The simplest thing to do is to wrap the input into a unified struct as high in the call chain as possible, and unpack it only  where it is needed (only in GramMatrix). The QN solver in cuML does the same: [pack inputs into a unified sturcture at the entry point](https://github.com/rapidsai/cuml/blob/6976c8b893c57b4a8ae9a4929d06cb2dc0dc4b5e/cpp/src/glm/qn/qn.cuh#L161-L200).

If you are against a unified sparse / dense type in raft, then we could keep that in cuML, but it I think we should think about having such interface in raft.

Reply:
From my understanding, the code duplication in `GramMatrix` is only at the API layer but the actual logic for each input types are separate pieces anyway. Here's my suggestion, since it's trivial to overload the API layer as that doesn't have significant impact to compile time or binary size and I don't think we need any template methods here. For `GramMatrix`:
1. Keep the dense API
2. Keep the sparse API
3. Introduce unified API. This is just a facade, we immediately dispatch to the above two

This way, we solve all issues. The cuML call chain is simplified, and for users of RAFT that may only care about either the sparse API or the dense API can call that directly without us forcing them to add another type layer of `unified` before interacting with RAFT.

Reply:
My apologies @tfeher and @mfoerste4, I meant to revisit this discussion myself and things have been pretty crazy the past couple of weeks. Please understand that my scrutiny over adding new types is only because 

1) I want to make sure that we're not exploding the number of types that a user needs work with, and 
2) making sure that we aren't exploding the number of new types each time we encounter a new pattern, unless said types can be at least moderately reusable across the codebase.

It sounds like the goal here is to have a unified base class. So far, the cython APIs in RAFT have gotten pretty tedious to code- and that's just for host/device mdspan. Adding sparse to the mid and accepting that on host and device is going to make things even more tedious. I definiely understand the reasoning here. 

Currently, the `SimpleMat` contains linear algebra implementations. If the goal here is just to have a unified base class that can accept read-only inputs, could we maybe just create a very simple class called something like `read_only_matrix_view` which we extend for `csr_read_only_matrix_view` and `dense_read_only_matrix_view`?

Reply:
> As for cuML, in addition to the read-only views we also need a data/structure-owning matrix wrapper that allows for resizing internal data structures ([ResizableCsrMatrix](https://github.com/mfoerste4/cuml/blob/078ba5dc744477e1a8c57c0799fe8af981e4c3a9/cpp/src/svm/sparse_util.cuh#L34)).

It sounds like what you need there is indeed a `structure-owning` CSR matrix. Why not use [the one](https://github.com/rapidsai/raft/blob/branch-23.04/cpp/include/raft/core/device_csr_matrix.hpp#L156) that's already in RAFT? Again, it looks like we're duplicating types, which means increased maintenance burden.